File: api/typescript-client/tests/storyboard.test.ts
import { describe, it, expect } from "vitest";
import { mkdtempSync, readFileSync } from "fs";
import { join } from "path";
import { tmpdir } from "os";
import { Storyboard } from "../src/index.js";

function makeTmpDir(): string {
  return mkdtempSync(join(tmpdir(), "storyboard-test-"));
}

describe("storyboard", () => {
  it("creates narration entry with text", async () => {
    const dir = makeTmpDir();
    const sb = new Storyboard(dir);
    await sb.beginNarration("Hello world");
    await sb.endNarration();

    const data = JSON.parse(readFileSync(join(dir, "storyboard.json"), "utf-8"));
    expect(data.narrations).toHaveLength(1);
    expect(data.narrations[0].narrationId).toBe(0);
    expect(data.narrations[0].text).toBe("Hello world");
  });

  it("auto-increments narration IDs", async () => {
    const dir = makeTmpDir();
    const sb = new Storyboard(dir);
    await sb.beginNarration("First");
    await sb.endNarration();
    await sb.beginNarration("Second");
    await sb.endNarration();

    const data = JSON.parse(readFileSync(join(dir, "storyboard.json"), "utf-8"));
    expect(data.narrations[0].narrationId).toBe(0);
    expect(data.narrations[1].narrationId).toBe(1);
  });

  it("stores voice on narration", async () => {
    const dir = makeTmpDir();
    const sb = new Storyboard(dir, undefined, { voices: { douglas: { en: "am_adam" } } });
    await sb.beginNarration("Hello", undefined, "douglas");
    await sb.endNarration();

    const data = JSON.parse(readFileSync(join(dir, "storyboard.json"), "utf-8"));
    expect(data.narrations[0].voice).toBe("douglas");
  });

  it("omits voice when not specified", async () => {
    const dir = makeTmpDir();
    const sb = new Storyboard(dir);
    await sb.beginNarration("Hello");
    await sb.endNarration();

    const data = JSON.parse(readFileSync(join(dir, "storyboard.json"), "utf-8"));
    expect(data.narrations[0].voice).toBeUndefined();
  });

  it("stores translations on narration", async () => {
    const dir = makeTmpDir();
    const sb = new Storyboard(dir);
    await sb.beginNarration("Hello", { de: "Hallo" });
    await sb.endNarration();

    const data = JSON.parse(readFileSync(join(dir, "storyboard.json"), "utf-8"));
    expect(data.narrations[0].translations.de).toBe("Hallo");
  });

  it("includes videoFile path on narration", async () => {
    const dir = makeTmpDir();
    const sb = new Storyboard(dir);
    await sb.beginNarration("Hello");
    await sb.endNarration();

    const data = JSON.parse(readFileSync(join(dir, "storyboard.json"), "utf-8"));
    expect(data.narrations[0].videoFile).toBe("videos/narration-000.mp4");
  });

  it("includes voices in options", async () => {
    const dir = makeTmpDir();
    const sb = new Storyboard(dir, undefined, {
      voices: { douglas: { en: "am_adam" }, natalie: { en: "bf_alice" } },
    });
    await sb.beginNarration("Hello", undefined, "douglas");
    await sb.endNarration();

    const data = JSON.parse(readFileSync(join(dir, "storyboard.json"), "utf-8"));
    expect(data.options.voices.douglas.en).toBe("am_adam");
    expect(data.options.voices.natalie.en).toBe("bf_alice");
  });

  it("throws when nested narrations", async () => {
    const dir = makeTmpDir();
    const sb = new Storyboard(dir);
    await sb.beginNarration("First");
    await expect(sb.beginNarration("Second")).rejects.toThrow(
      "Cannot begin a new narration while another is still open"
    );
  });

  it("done succeeds after all narrations closed", async () => {
    const dir = makeTmpDir();
    const sb = new Storyboard(dir);
    await sb.beginNarration("Hello");
    await sb.endNarration();
    await expect(sb.done()).resolves.toBeUndefined();
  });

  it("done throws if narration is open", async () => {
    const dir = makeTmpDir();
    const sb = new Storyboard(dir);
    await sb.beginNarration("Hello");
    await expect(sb.done()).rejects.toThrow("Cannot finalize: a narration bracket is still open");
  });
});

File: api/java-client/src/test/java/screencastnarrator/StoryboardTest.java
package screencastnarrator;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

class StoryboardTest {

    private final ObjectMapper mapper = new ObjectMapper();

    private JsonNode readStoryboard(Path dir) throws Exception {
        return mapper.readTree(Files.readString(dir.resolve("storyboard.json")));
    }

    @Test
    void createsNarrationEntryWithText(@TempDir Path dir) throws Exception {
        Storyboard sb = new Storyboard(dir);
        sb.beginNarration("Hello world");
        sb.endNarration();

        JsonNode data = readStoryboard(dir);
        JsonNode narrations = data.get("narrations");
        assertEquals(1, narrations.size());
        assertEquals(0, narrations.get(0).get("narrationId").asInt());
        assertEquals("Hello world", narrations.get(0).get("text").asText());
    }

    @Test
    void autoIncrementsNarrationIds(@TempDir Path dir) throws Exception {
        Storyboard sb = new Storyboard(dir);
        sb.beginNarration("First");
        sb.endNarration();
        sb.beginNarration("Second");
        sb.endNarration();

        JsonNode narrations = readStoryboard(dir).get("narrations");
        assertEquals(0, narrations.get(0).get("narrationId").asInt());
        assertEquals(1, narrations.get(1).get("narrationId").asInt());
    }

    @Test
    void storesVoiceOnNarration(@TempDir Path dir) throws Exception {
        Storyboard sb = new Storyboard(dir, null, "en", null, false, 24,
                Map.of("douglas", Map.of("en", "am_adam")));
        sb.beginNarration("Hello", Map.of(), "douglas");
        sb.endNarration();

        JsonNode narration = readStoryboard(dir).get("narrations").get(0);
        assertEquals("douglas", narration.get("voice").asText());
    }

    @Test
    void omitsVoiceWhenNotSpecified(@TempDir Path dir) throws Exception {
        Storyboard sb = new Storyboard(dir);
        sb.beginNarration("Hello");
        sb.endNarration();

        JsonNode narration = readStoryboard(dir).get("narrations").get(0);
        assertNull(narration.get("voice"));
    }

    @Test
    void storesTranslationsOnNarration(@TempDir Path dir) throws Exception {
        Storyboard sb = new Storyboard(dir);
        sb.beginNarration("Hello", Map.of("de", "Hallo"));
        sb.endNarration();

        JsonNode narration = readStoryboard(dir).get("narrations").get(0);
        assertEquals("Hallo", narration.get("translations").get("de").asText());
    }

    @Test
    void includesVideoFilePathOnNarration(@TempDir Path dir) throws Exception {
        Storyboard sb = new Storyboard(dir);
        sb.beginNarration("Hello");
        sb.endNarration();

        JsonNode narration = readStoryboard(dir).get("narrations").get(0);
        assertEquals("videos/narration-000.mp4", narration.get("videoFile").asText());
    }

    @Test
    void includesVoicesInOptions(@TempDir Path dir) throws Exception {
        Storyboard sb = new Storyboard(dir, null, "en", null, false, 24,
                Map.of("douglas", Map.of("en", "am_adam"),
                       "natalie", Map.of("en", "bf_alice")));
        sb.beginNarration("Hello", Map.of(), "douglas");
        sb.endNarration();

        JsonNode options = readStoryboard(dir).get("options");
        assertEquals("am_adam", options.get("voices").get("douglas").get("en").asText());
        assertEquals("bf_alice", options.get("voices").get("natalie").get("en").asText());
    }

    @Test
    void throwsWhenNestedNarrations(@TempDir Path dir) throws Exception {
        Storyboard sb = new Storyboard(dir);
        sb.beginNarration("First");

        assertThrows(IllegalStateException.class, () -> sb.beginNarration("Second"));
    }

    @Test
    void doneSucceedsAfterAllNarrationsClosed(@TempDir Path dir) throws Exception {
        Storyboard sb = new Storyboard(dir);
        sb.beginNarration("Hello");
        sb.endNarration();

        assertDoesNotThrow(sb::done);
    }

    @Test
    void doneThrowsIfNarrationIsOpen(@TempDir Path dir) throws Exception {
        Storyboard sb = new Storyboard(dir);
        sb.beginNarration("Hello");

        assertThrows(IllegalStateException.class, sb::done);
    }

    @Test
    void debugOverlayAndFontSizeInOptions(@TempDir Path dir) throws Exception {
        Storyboard sb = new Storyboard(dir, null, "en", null, true, 48);
        sb.beginNarration("Test");
        sb.endNarration();

        JsonNode options = readStoryboard(dir).get("options");
        assertTrue(options.get("debugOverlay").asBoolean());
        assertEquals(48, options.get("fontSize").asInt());
    }

    @Test
    void noOptionsWhenDefaults(@TempDir Path dir) throws Exception {
        Storyboard sb = new Storyboard(dir);
        sb.beginNarration("Test");
        sb.endNarration();

        assertNull(readStoryboard(dir).get("options"));
    }
}
